src/System.Linq/src/System/Linq/Lookup.cs

```diff
@@ -247,8 +245,7 @@ public IEnumerable<TResult> ApplyResultSelector<TResult>(Func<TKey, IEnumerable<
                 do
                 {
                     g = g._next;
-                    g.Trim();
-                    yield return resultSelector(g._key, g._elements);
+                    yield return resultSelector(g._key, g);
```

Comment:
Another reason might be to separate keys from elements before passing to the selector - that is just to ensure if eventually only elements are retained, they do not hold the keys as well.